rfc: propose split op for oneDNN Graph API

xiang1guo opened the pull request:
This RFC is to add a new Split operation to oneDNN graph API opset.

The rendered version: [link](https://github.com/xiang1guo/oneDNN/tree/xiang/rfcs/graph-split-op/rfcs/20230323-add-split-graph-op)

xiang1guo commented:
Can you help to review this RFC? Thanks! @igorsafo @mgouicem @dzarukin

chunyuan-w commented:
1. PyTorch [split OP](https://pytorch.org/docs/stable/generated/torch.split.html) has a `split_size_or_sections` input:
- When it is `list`, it is aligned with the `size_splits` attr here.
- When it is an `int`, it seems that there's no correspondent attr:
   - PyTorch `int` `split_size_or_sections`:
      - It is the size of the chunk and it denotes that the input will be split into equally sized chunks (if possible).
      - Last chunk will be smaller if the tensor size along the given dimension dim is not divisible by `split_size`. 

   - `num_splits` in oneDNN graph:
      -  It is the number of chunks
      - The dimension of input tensor data shape along `axis` must be evenly divisible by `num_splits` attribute.
 
In static shape cases, if the input shape is known, the integration could try to convert `int` `split_size_or_sections` to `num_splits` in case the dimension could be evenly divisible by `num_splits`. In dynamic shape cases, I feel like it will be hard for the integration to map the OP to oneDNN graph.

2. For PyTorch, each of the output chunk is a view of the original tensor. If it is not the case in oneDNN graph, may I know if this will be mentioned anywhere in the document?

sanchitintel commented:
@xiang1guo, we plan to emulate dynamic shape support in IPEX with multiple compiled partitions for various static shapes.
Since the `num_splits` attribute is to be provided at the time of graph creation, IPEX would provide its value for the first input but then it might change for a later input! :(

cc @chunyuan-w

xiang1guo commented:
> @xiang1guo, we plan to emulate dynamic shape support in IPEX with multiple compiled partitions for various static shapes. Since the `num_splits` attribute is to be provided at the time of graph creation, IPEX would provide its value for the first input but then it might change for a later input! :(
> 
> cc @chunyuan-w

Do you mean that 

> 1. PyTorch [split OP](https://pytorch.org/docs/stable/generated/torch.split.html) has a `split_size_or_sections` input:
> 
> * When it is `list`, it is aligned with the `size_splits` attr here.
> * When it is an `int`, it seems that there's no correspondent attr:
>   
>   * PyTorch `int` `split_size_or_sections`:
>     
>     * It is the size of the chunk and it denotes that the input will be split into equally sized chunks (if possible).
>     * Last chunk will be smaller if the tensor size along the given dimension dim is not divisible by `split_size`.
>   * `num_splits` in oneDNN graph:
>     
>     * It is the number of chunks
>     * The dimension of input tensor data shape along `axis` must be evenly divisible by `num_splits` attribute.
> 
> In static shape cases, if the input shape is known, the integration could try to convert `int` `split_size_or_sections` to `num_splits` in case the dimension could be evenly divisible by `num_splits`. In dynamic shape cases, I feel like it will be hard for the integration to map the OP to oneDNN graph.
> 
> 2. For PyTorch, each of the output chunk is a view of the original tensor. If it is not the case in oneDNN graph, may I know if this will be mentioned anywhere in the document?

For 1, as you mentioned in the meeting, we can add another attribute to align the meaning of Pytorch. It helps for mapping, but it still just can't handle the situation when the `axis` dimension of the input tensor is dynamic. @ZhennanQin @yifeizh2 Do you have any comments?

For 2, yes, graph API doesn't guarantee the view or copy. I will add more descriptions for it. Thanks for your reminder!

ZhennanQin commented:
> For 1, as you mentioned in the meeting, we can add another attribute to align the meaning of Pytorch. It helps for mapping, but it still just can't handle the situation when the axis dimension of the input tensor is dynamic.

What do you mean by `the axis dimension of the input tensor is dynamic`? Can you provide an example for better understanding?

xiang1guo commented:
> > For 1, as you mentioned in the meeting, we can add another attribute to align the meaning of Pytorch. It helps for mapping, but it still just can't handle the situation when the axis dimension of the input tensor is dynamic.
> 
> What do you mean by `the axis dimension of the input tensor is dynamic`? Can you provide an example for better understanding?

For dynamic input data [1, 28, ?](? means this dimension can be gotten in the execution stage) and the `axis=2`.

ZhennanQin commented:
> > > For 1, as you mentioned in the meeting, we can add another attribute to align the meaning of Pytorch. It helps for mapping, but it still just can't handle the situation when the axis dimension of the input tensor is dynamic.
> > 
> > 
> > What do you mean by `the axis dimension of the input tensor is dynamic`? Can you provide an example for better understanding?
> 
> For dynamic input data [1, 28, ?](? means this dimension can be gotten in the execution stage) and the `axis=2`.

I see. Thanks for the example. In this case, the number of outputs from `split` is dynamic, how's this represented in the framework graph?

xiang1guo commented:
>

xiang1guo commented:
> > > > For 1, as you mentioned in the meeting, we can add another attribute to align the meaning of Pytorch. It helps for mapping, but it still just can't handle the situation when the axis dimension of the input tensor is dynamic.
> > > 
> > > 
> > > What do you mean by `the axis dimension of the input tensor is dynamic`? Can you provide an example for better understanding?
> > 
> > 
> > For dynamic input data [1, 28, ?](? means this dimension can be gotten in the execution stage) and the `axis=2`.
> 
> I see. Thanks for the example. In this case, the number of outputs from `split` is dynamic, how's this represented in the framework graph?

Seems for now we don't have any real case. @chunyuan-w @sanchitintel  Do you have any real case about it?

wuxun-zhang commented:
> @xiang1guo, we plan to emulate dynamic shape support in IPEX with multiple compiled partitions for various static shapes. Since the `num_splits` attribute is to be provided at the time of graph creation, IPEX would provide its value for the first input but then it might change for a later input! :(
> 
> cc @chunyuan-w

Hi @sanchitintel, by saying `then it might change for a later input`, do you mean that `num_splits` will be no longer an attribute? How is it handled in bridge?

sanchitintel commented:
> by saying `then it might change for a later input`, do you mean that `num_splits` will be no longer an attribute? How is it handled in bridge?

Hi @wuxun-zhang, if the input-size would change, then `num_splits` may change as well, but since the framework would have passed a value of `num_splits` corresponding to the first input encountered at the time of graph creation, I think we would run into a shape-mismatch issue at runtime.

So, it'd be great if we could pass this value at compile time instead. IIRC, something similar is done for DynamicReshape as well.

sanchitintel commented:
> In this case, the number of outputs from `split` is dynamic, how's this represented in the framework graph?

@ZhennanQin, currently, the IPEX/PyTorch JIT framework graph is created by tracing with a sample input, so the framework graph's nodes are specific to that input, so, for the case you mentioned, another input with a different input-shape would have generated a different graph. So, if we were to lower the framework graph to oneDNN Graph for an input that would generate a different number of tensors for some op, then we would encounter a runtime error.

@frost-intel @chunyuan-w, do you know how Dynamo handles such cases? Thanks!

frost-intel commented:
> @frost-intel @chunyuan-w, do you know how Dynamo handles such cases? Thanks!

Dynamo can either use a constant input size similar to PT JIT, or it can have a dynamic size marked as a SymInt object with undetermined size. We haven't explored dynamic sizes in our Dynamo/oneDNN Graph bridge, but an attribute or flag marking a dimension as Dynamic would likely be required.

TaoLv commented on rfcs/20230323-add-split-graph-op/README.md:
```diff
@@ -0,0 +1,152 @@
+ optimization is limited. So we propose to add the `Split` op into graph API.
+
+
+## Proposal
```
Before proposing options for graph API, I feel it would be better to add a section describing how the op and functionality are implemented in mainstream frameworks and toolkits.

xiang1guo replied:
@TaoLv  Done for that, please help to check

xiang1guo commented:
@chunyuan-w @sanchitintel @retonym @ZhennanQin I just updated the RFC and added an **option2** per pytorch team's request. Please help to review the **option2** again, if confirmed, our final implementation will be based on this option.

ZhennanQin commented on rfcs/20230323-add-split-graph-op/README.md:
```diff
@@ -0,0 +1,161 @@
+| input     | `src`         | input data                                                                             | Required             | f32,f16,bf16* |
+| attribute | `axis`        | Specifies which dimension to split along.                                              | Required             | s64           |
+| attribute | `num_splits`  | number of outputs into which the input tensor data will be split along axis dimension. | optional             | s64           |
+| attribute | `size_split`  | size for single output tensor along axis dimension.                                    | optional             | s64           |
```
during graph build stage, what if `size_split` is specified while input shape is unknown? How to determine the output number?

xiang1guo replied:
That case you mentioned is a dynamic case that is not supported and has not been encountered yet. Once the dynamic shape feature lands in the future release, we can take it into consideration. cc@wuxun-zhang. 
The option2 here is for the convenience of integration on the IPEX side which can simply map the `size_split_or_sections` into `size_split` during the build stage.

ZhennanQin replied:
I think it's not a dynamic shape case. FWK won't provide concrete shapes during the graph build stage even for models with static shapes.

wuxun-zhang replied:
> during graph build stage, what if `size_split` is specified while input shape is unknown? How to determine the output number?

Why need to determine the output number at graph build stage?

ZhennanQin replied:
> Why need to determine the output number at graph build stage?

If you don't know the output number, how to build the graph after this op? A simple example is split+concat, how to collect all the outputs from split and feed them into contact op?

wuxun-zhang replied:
I see, it's related to how bridge pass framework's graph to oneDNN Graph. We may need framework team to help answer, cc @chunyuan-w @sanchitintel @retonym.

chunyuan-w replied:
Do I understand correctly that this is a case where the graph structure depends on the input shapes? In the current integration, the graph will only be built once. I feel like it does not work anymore in such cases. Let me check with @jgong5 & @sanchitintel on it.

sanchitintel replied:
@chunyuan-w, I agree that this case would encounter a runtime error with our multiple compiled partitions per partition approach. I think we can't solve this problem without re-creating the graph.

yifeizh2 commented on rfcs/20230323-add-split-graph-op/README.md:
```diff
@@ -0,0 +1,161 @@
+- Uniformly specify the size of each output chunk through an integer
+- Specify the specific size of each chunk through a list. 
+
+The size of all output chunks in the first two rules is the same but may differ 
```
**Rule 2** above may lead to a tail chunk having a smaller shape than the specified shape (as was for the case of PyTorch and also specified in option 2), so please revise the statement here.

xiang1guo replied:
ok, will do it.

wuxun-zhang commented on rfcs/20230323-add-split-graph-op/README.md:
```diff
@@ -0,0 +1,161 @@
+| attribute | `axis`        | Specifies which dimension to split along.                                              | Required             | s64           |
+| attribute | `num_splits`  | number of outputs into which the input tensor data will be split along axis dimension. | optional             | s64           |
+| attribute | `size_split`  | size for single output tensor along axis dimension.                                    | optional             | s64           |
+| attribute | `size_splits` | list of sizes for each output alone axis dimension.                                    | optional             | s64           |
```
The naming of `size_splits` and `size_split` are much similar, but they have very different semantics as you note below. It may be confuse users, suggest to give a more descriptive naming.

xiang1guo replied:
ok, what about `sections_split` just similar to PyTorch.

Jianhui-Li commented on rfcs/20230323-add-split-graph-op/README.md:
```diff
@@ -0,0 +1,161 @@
+
+#### operations used in Frameworks and toolkit
+
+| Tensorflow | Argument Name         | Description                                                                                       |
```
As TF workloads moves to XLA compiler path, and Pytorch move to Torchdynamic/TorchInductor, FW operations are decomposed to XLA HLO op or Pytorch PrimTorch OP.  These operations usually has simpler semantics.  oneDNN Graph operation should first consider what is the basic operation we need to support, and usually these operation design would give better insight - what operations are popular (those unpopular ones may not lowered to these simple ops), and how the basic operations look like. 

Could you please also include how these two set operations support split?

xiang1guo replied:
Hi, @Jianhui-Li, just checked the XLA and PrimTorch, these two operations don't directly support `Split`. 
- TensorFlow's `Split` will be lowered to XLA HLO [Slice op](https://www.tensorflow.org/xla/operation_semantics#slice). I have updated the definition of XLA HLO `Slice` in this section.  
- For Pytorch/PrimTorch, when Torch IR is lowered to Aten IR/Prim IR, the `torch.Split` will call `aten.Split` and no decomposition to PrimTorch ops happened during this stage. And inductor will lower `aten.Split` to `slice` in its implementation.  cc @retonym @chunyuan-w

TaoLv commented:
The proposal was reviewed internally and will not be implemented in the library due to the complexity of defining a split op to support different usages in different frameworks. Thank you all for the discussion and participation.
